### PR TITLE
feat: NFT avatar

### PR DESCRIPTION
**Pull request by lelenei**
closes #

### CODE REVIEW

File: packages/maskbook/src/utils/messages.ts [hunk 1]
```diff
@@ -77,6 +85,8 @@ export interface MaskMessages extends SettingsEvents {
         requestID: string
         selectedPersona: PersonaIdentifier
     }
+
+    NFTAvatorUpdated: NFTAVatorEvent
```
Comment (guanbinrui) [inline packages/maskbook/src/utils/messages.ts:89]:
```suggestion
    NFTAvatarUpdated: NFTAVatarEvent
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/ui-provider.ts [hunk 2]
```diff
@@ -160,6 +162,8 @@ const twitterUI: SocialNetworkUI.Definition = {
         setupWizard: createTaskStartSetupGuideDefault('twitter.com'),
         userBadge: injectMaskUserBadgeAtTwitter,
         commentComposition: undefined,
+        userAvator: injectNFTAvatorInTwitter,
+        enhancedProfileNFTAvator: injectProfileNFTAvatorInTwitter,
```
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/ui-provider.ts:166]:
```suggestion
        userAvator: injectNFTAvatarInTwitter,
        enhancedProfileNFTAvator: injectProfileNFTAvatarInTwitter,
```

File: packages/maskbook/src/components/InjectedComponents/NFTAvator.tsx [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            <Button className={classes.button} variant="text" onClick={() => retry()}>
+                {t('plugin_collectible_retry')}
+            </Button>
+        </Box>
+    )
+    return (
+        <>
+            <Box className={classes.root}>
+                <Box className={classes.title}>
+                    <Typography variant="body1">NFT Avator Setting</Typography>
```
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/NFTAvator.tsx:148]:
```suggestion
                    <Typography variant="body1" color="textPrimary">NFT Avator Setting</Typography>
```

File: packages/maskbook/src/components/InjectedComponents/NFTAvator.tsx [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                    {account ? (
+                        <Typography variant="body1">
```
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/NFTAvator.tsx:150]:
```suggestion
                        <Typography variant="body1" color="textPrimary">
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/Avatar.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        width: '100%',
+        paddingLeft: 10,
+    },
+    text: {
+        fontSize: 10,
+        transform: 'scale(0.8)',
+        margin: 0,
+        color: 'white',
+        whiteSpace: 'nowrap',
+        textShadow: '2px 1px black',
```
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/Avatar.tsx:41]:
```
textShadow: '1px 1px black, 1px 0px black, 0px 1px black, -1px 0px black, 0px -1px black, -1px -1px black, 1px -1px black, -1px 1px black',
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/NFTAvatarInTwitter.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+export function injectNFTAvatarInTwitter(signal: AbortSignal) {
+    const watcher = new MutationObserverWatcher(searchTwitterAvatarSelector())
+    startWatch(watcher, signal)
+    createReactRootShadowed(watcher.firstDOMProxy.afterShadow, { signal }).render(<NFTAvatarInTwitter />)
+}
+
+const useStyles = makeStyles()((theme) => ({
+    root: {
+        position: 'absolute',
+        top: 12,
```
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/NFTAvatarInTwitter.tsx:27]:
```suggestion
        top: 18,
```

File: packages/dashboard/src/pages/CreateMaskWallet/en.html [hunk 7]
```diff
@@ -4,7 +4,8 @@
         <meta charset="UTF-8" />
         <meta
             name="viewport"
-            content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0" />
+            content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0"
+        />
```
Comment (UncleBill) [inline packages/dashboard/src/pages/CreateMaskWallet/en.html:8]: You need to upgrade you local prettier to `2.4.0`, and format files with it to revert this kind of change.

File: packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        },
+        loading,
+        retry,
+        error,
+    } = useCollectibles(account, chainId, provider, page, 50)
+    //0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
+
+    const { collectibles, hasNextPage } = value
+
+    console.log(collectibles)
```
Comment (zhouhanseng) [inline packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx:102]: remove debug code

File: packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    useEffect(() => {
+        setCollectibles_(collectibles)
+    }, [collectibles, collectibles.length])
```
Comment (zhouhanseng) [inline packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx:105]: why dependent on `collectibles` and `collectibles.length`

File: packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+    const onClick = useCallback(async () => {
+        if (!selectedToken) return
+        onChange(selectedToken)
+    }, [onChange, selectedToken])
+
+    const onAddClick = useCallback((token) => {
+        setSelectedToken(token)
+        setCollectibles_((tokens) => uniqBy([token, ...tokens], (x) => x.contractDetailed.address && x.tokenId))
+    }, [])
```
Comment (zhouhanseng) [inline packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx:115]: Add `tokens` to the dependency list?

File: packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+    const { openDialog: openSelectProviderDialog } = useRemoteControlledDialog(
+        WalletMessages.events.selectProviderDialogUpdated,
+    )
+
+    const LoadStatus = Array.from({ length: 24 })
+        .fill(0)
+        .map((_, i) => <Skeleton animation="wave" variant="rectangular" className={classes.image} key={i} />)
```
Comment (zhouhanseng) [inline packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx:123]: Dose A Loading UI contains 24 same styling `<Skeleton />` looks okay?

File: packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+                            />
+                        ) : null}
+                        <Box className={classes.button}>
+                            <Button
+                                variant="contained"
+                                size="medium"
+                                className={classes.setNFTAvatar}
+                                onClick={() => onClick()}
+                                disabled={!selectedToken}>
+                                Set NFT Avatar
```
Comment (zhouhanseng) [inline packages/maskbook/src/components/InjectedComponents/NFTAvatar.tsx:202]: Add to locale kit？

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/profileNFTAvatar.tsx [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    createReactRootShadowed(watcher.firstDOMProxy.afterShadow, { signal }).render(<NFTAvatarInTwitter />)
+}
+
+const useStyles = makeStyles()((theme) => ({
+    root: {
+        padding: '11px 14px 11px 14px',
+    },
+}))
+
+function useGetCurrentUserInfo(): { userId?: string; identifier?: ProfileIdentifier } | undefined {
```
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/profileNFTAvatar.tsx:30]:
```suggestion
function useCurrentUserInfo(): { userId?: string; identifier?: ProfileIdentifier } | undefined {
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/Avatar.tsx [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    avatar: AvatarMetaDB
+}
+function NFTBadgeInTwitter({ avatar }: NFTBadgeInTwitterProps) {
+    const { classes } = useStyles()
+    return (
+        <div
+            className={classes.root}
+            onClick={(e) => {
+                e.preventDefault()
+                window.open(resolveOpenSeaLink(avatar.address, avatar.tokenId), '_blank')
```
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/Avatar.tsx:57]: Remove open in JavaScript. Please use `<Link />` instead.

File: packages/maskbook/src/extension/options-page/DashboardRouters/Settings.tsx [hunk 15]
```diff
@@ -284,6 +287,23 @@ export default function DashboardSettingsRouter() {
                         </Card>
                     </Paper>
 
+                    <Paper component="section" className={classes.section} elevation={elevation}>
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardRouters/Settings.tsx:290]: It should only be available in the insider or beta versions.

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/NFTAvatarInTwitter.tsx [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+        <>
+            {avatarId === avatar.avatarId ? (
+                <div className={classes.root}>
+                    <div className={classes.nftLogo}>
+                        <NFTAvatarAmountIcon className={classes.nftImage} />
+                    </div>
+                    <div className={classes.wrapper}>
+                        <div className={classes.amountWrapper}>
+                            <Typography align="center" className={classes.amount}>
+                                {`${amount} ETH`}
```
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/NFTAvatarInTwitter.tsx:139]: Are you sure you have formatted the amount properly? Test edge cases like `0.000000001` or `100000000.0000000001`.

File: packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+                    ? []
+                    : asset.buyOrders
+                : asset.orders
+            : asset.sellOrders
+
+    const order = head(
+        orders.sort(
+            (a, b) =>
+                new BigNumber(getOrderUnitPrice(a) ?? 0).toNumber() -
+                new BigNumber(getOrderUnitPrice(b) ?? 0).toNumber(),
```
Comment (UncleBill) [inline packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx:241]: Compare with `Number` method `isGreaterThan` or `gt`

File: packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+export async function saveNFTAvatar(userId: string, avatarId: string, contract: string, tokenId: string) {
+    const asset = await PluginCollectibleRPC.getAsset(contract, tokenId)
+    const orders =
+        !asset.sellOrders || asset.sellOrders?.length === 0
+            ? !asset.orders || asset.orders?.length === 0
+                ? !asset.buyOrders || asset.buyOrders?.length === 0
+                    ? []
+                    : asset.buyOrders
+                : asset.orders
+            : asset.sellOrders
```
Comment (UncleBill) [inline packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx:235]: It's hard to understand at first glance, also hard to refactor or modify in the future. Try following code:

```ts
let orders = []
if (asset.sellOrders?.length) {
  orders = asset.sellOrders
} else if (asset.buyOrders?.length) {
  orders = asset.buyOrders
}
```

File: packages/maskbook/src/components/InjectedComponents/NFT/NFTBadge.tsx [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+                title={resolveOpenSeaLink(avatar.address, avatar.tokenId)}
+                href={resolveOpenSeaLink(avatar.address, avatar.tokenId)}
+                target="_blank"
+                rel="noopener noreferrer">
+                <NFTAvatarAmountIcon className={classes.icon} />
+                <div className={classes.wrapper}>
+                    <Typography className={classes.text}>{`${avatar.amount} ETH`}</Typography>
+                </div>
+            </Link>
+        </div>
```
Comment (UncleBill) [inline packages/maskbook/src/components/InjectedComponents/NFT/NFTBadge.tsx:72]: You already have a `<Link />` inside.

File: packages/maskbook/src/extension/options-page/DashboardComponents/SearchInput.tsx [hunk 20]
```diff
@@ -1,65 +1,7 @@
...
-                        setVisible(e.target.value.length !== 0)
-                        onChange?.(e.target.value)
-                    }}
-                />
-            </Paper>
-        </Paper>
-    )
+export interface SearchInputProps extends InputBoxProps {}
+export function SearchInput(props: SearchInputProps) {
+    return <InputBox {...props} children={<SearchIcon />} />
```
Comment (UncleBill) [inline packages/maskbook/src/extension/options-page/DashboardComponents/SearchInput.tsx:6]:
```suggestion
    return <InputBox children={<SearchIcon />} {...props} />
```

Then it's possible to override `<SearchIcon />` from outside.

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/NFT/NFTAvatarInTwitter.tsx [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+                })
+        },
+        [enqueueSnackbar],
+    )
+
+    useEffect(() => {
+        return MaskMessage.events.NFTAvatarUpdated.on((data) => {
+            onUpdate(data)
+        })
+    }, [])
```
Comment (UncleBill) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/NFT/NFTAvatarInTwitter.tsx:83]:
```suggestion
    }, [onUpdate])
```

File: packages/maskbook/src/components/InjectedComponents/NFT/NFTBadge.tsx (hunk 19 above)
Comment (lelenei) [reply packages/maskbook/src/components/InjectedComponents/NFT/NFTBadge.tsx:72]: Link cannot jump

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/Avatar.tsx (hunk 14 above)
Comment (lelenei) [reply packages/maskbook/src/social-network-adaptor/twitter.com/injection/Avatar.tsx]: Link cannot jump

File: packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx (hunk 17 above)
Comment (lelenei) [reply packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx:241]: isGreaterThan  returns  bool

Comment (UncleBill) [reply packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx:241]: > isGreaterThan returns bool

Then you should `BigNumber#minus` and `BigNumber#toNumber`

Comment (lelenei) [reply packages/maskbook/src/components/InjectedComponents/NFT/NFTAvatar.tsx:243]: minus is ok

**Review by guanbinrui — APPROVED** [review]
